rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  through flake inputs or `fetchX` builtins.
+
+Pure mode can be disabled by passing `--impure` on the command line.
+
+## Evaluation caching
+
+The fact that evaluation is now hermetic allows `nix` to cache flake
+attributes. For example (doing `nix build` on an already present
+package):
+
+    $ time nix build nixpkgs:firefox
+    real    0m1.497s
+
+    $ time nix build nixpkgs:firefox
+    real    0m0.052s
+
+The evaluation cache is kept in `~/.cache/nix/eval-cache-v1.sqlite`,
+which has entries like
+
+    INSERT INTO Attributes VALUES(
+      X'92a907d4efe933af2a46959b082cdff176aa5bfeb47a98fabd234809a67ab195',
+      'packages.firefox',
+      1,
+      '/nix/store/pbalzf8x19hckr8cwdv62rd6g0lqgc38-firefox-67.0.drv /nix/store/g6q0gx0v6xvdnizp8lrcw7c4gdkzana0-firefox-67.0 out');
+
+where the hash `92a9...` is a fingerprint over the flake store path
+and the contents of its lockfile. Because flakes are evaluated in pure
+mode, this uniquely identifies the evaluation result.
+
+Currently caching is only done for top-level attributes (e.g. for
+`packages.firefox` in the command above). In the future, we could also
+add other evaluated values to the cache (e.g. `packages.stdenv`) to
+speed up subsequent evaluations of other top-level attributes.
+
+## Flake-related commands
+
+The command `nix flake` has various subcommands for managing
+flakes. These are:
+
+* `nix flake list`: Show all flakes in the global and local registry.
+
+* `nix flake add <from-flake-ref> <to-flake-ref>`: Add an entry to the
+  local registry. (Recall that this overrides the global registry.)
+  For example,
+
+      # nix flake add nixpkgs github:my-repo/my-nixpkgs
+
+  redirects `nixpkgs` to a different repository. Similarly,
+
+      # nix flake add nixpkgs github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757
+
+  pins `nixpkgs` to a specific revision.
+
+  Note that registries only have an effect on flake references used on
+  the command line or when lock files are generated or updated.
+
+* `nix flake remove <flake-ref>`: Remove an entry from the local
+  registry.
+
+* `nix flake pin <flake-ref>`: Compute a locked flake reference and
+  add it to the local registry. For example:
+
+      # nix flake pin nixpkgs
+
+  will add a mapping like `nixpkgs` ->
+  `github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757` to
+  the local registry.
+
+* `nix flake init`: Create a skeleton `flake.nix` in the current
+  directory.
+
+* `nix flake update`: Recreate the lock file from scratch.
+
+* `nix flake check`: Do some checks on the flake (e.g. check that all
+  `packages` are really packages), then build the `checks`
+  derivations. For example, `nix flake check patchelf` fetches the
+  `patchelf` flake, evaluates it and builds it.
+
+* `nix flake clone`: Do a `git clone` to obtain the source of the
+  specified flake, e.g. `nix flake clone dwarffs` will yield a clone
+  of the `dwarffs` repository.
```

Comment:
And alternatively, once the clone is performed, does the lockfile point to the clone?

Reply:
The lockfile never points to its own repository (that would be impossible because changing the lockfile invalidates the repo hash). I.e. the lockfile only locks the dependencies, not the top-level flake.

If the flake is in a subdirectory, it will still clone the entire repository. This is necessary because the flake can refer to files in parent directories in the repository.